torch/onnx/_internal/exporter.py

```diff
@@ -0,0 +1,233 @@
+    Any,
+    Callable,
+    cast,
+    Final,
+    Mapping,
+    Optional,
+    Protocol,
+    runtime_checkable,
+    Sequence,
+    TYPE_CHECKING,
+    TypeVar,
+    Union,
+)
+
+import torch
+
+from torch.onnx._internal import _beartype
+
+# We can only import onnx from this module in a type-checking context to ensure that
+# 'import torch.onnx' continues to work without having 'onnx' installed. We fully
+# 'import onnx' inside of dynamo_export (by way of _assert_dependencies).
+if TYPE_CHECKING:
+    import onnx
+
+_DEFAULT_OPSET_VERSION: Final[int] = 18
+
+
+@_beartype.beartype
+class ExportOptions:
+    """Options to influence the TorchDynamo ONNX exporter."""
+
+    opset_version: Optional[int] = None
+    """The ONNX opset version the exporter should target. Defaults to the latest
+    supported ONNX opset version. The default version will increment over time as
+    ONNX continues to evolve."""
+
+    dynamic_shapes: Optional[bool] = None
+    """Shape information hint for input/output tensors.
+
+    - ``None``: the exporter determines the most compatible setting.
+    - ``True``: all input shapes are considered dynamic.
+    - ``False``: all input shapes are considered static."""
+
+    logger: Optional[logging.Logger] = None
+    """The logger for the ONNX exporter to use. Defaults to creating a child
+    logger named "torch.onnx" under the current logger (as returned by
+    :py:meth:`logging.getLogger`)."""
+
+    def __init__(
+        self,
+        *,
+        opset_version: Optional[int] = None,
+        dynamic_shapes: Optional[bool] = None,
+        logger: Optional[logging.Logger] = None,
+    ):
+        self.opset_version = opset_version
+        self.dynamic_shapes = dynamic_shapes
+        self.logger = logger
+
+
+@runtime_checkable
+class ExportOutputSerializer(Protocol):
+    """Protocol for serializing an ONNX graph into a specific format (e.g. Protobuf)."""
+
+    def serialize(
+        self, export_output: ExportOutput, destination: io.BufferedIOBase
+    ) -> None:
+        pass
+
+
+class ProtobufExportOutputSerializer:
+    """Serializes ONNX graph as Protobuf."""
+
+    @_beartype.beartype
+    def serialize(
+        self, export_output: ExportOutput, destination: io.BufferedIOBase
+    ) -> None:
+        import onnx
+
+        if not isinstance(export_output.model_proto, onnx.ModelProto):
+            raise ValueError("export_output.ModelProto is not an onnx.ModelProto")
+        destination.write(export_output.model_proto.SerializeToString())
+
+
+class ExportOutput:
+    """An in-memory representation of a PyTorch model that has been exported to ONNX."""
+
+    _export_options: Final[ExportOptions]
+    _model_proto: Final[onnx.ModelProto]
+
+    @_beartype.beartype
+    def __init__(self, export_options: ExportOptions, model_proto: onnx.ModelProto):
+        self._export_options = export_options
+        self._model_proto = model_proto
+
+    @property  # type: ignore[misc]
+    @_beartype.beartype
```

Comment:
@BowenBao what's the purpose of applying beartype here? There's nothing to validate?

Reply:
Is it worth actually validating `self`?

Reply:
More on validating return types.